orchagent/dash/dashpavalidationorch.cpp

```diff
@@ -0,0 +1,364 @@
+        return false;
+    }
+
+    auto& vni_addresses = vni->second;
+    auto address = vni_addresses.find(entry.address);
+    if (address == vni_addresses.end())
+    {
+        SWSS_LOG_WARN("PA Validaion entry VNI %u IP %s is removed or not created yet", entry.vni, entry.address.to_string().c_str());
+        return false;
+    }
+
+    SWSS_LOG_DEBUG("PA Validation entry (%u, %s) refcount is decreased to %" PRIu64, entry.vni, entry.address.to_string().c_str(), address->second - 1);
+
+    if (address->second == 1)
+    {
+        vni_addresses.erase(address);
+        if (vni_addresses.empty())
+        {
+            vni_map_.erase(entry.vni);
+        }
+
+        return true;
+    } else {
+        address->second--;
+        return false;
+    }
+}
+
+bool DashPaValidationOrch::fetchVniAddresses(uint32_t vni, PaValidationEntryAddresses& addresses) const
+{
+    SWSS_LOG_ENTER();
+
+    auto it = vni_map_.find(vni);
+    if (it == vni_map_.end())
+    {
+        return true;
+    }
+
+    const auto& vni_addresses = it->second;
+
+    addresses.reserve(vni_addresses.size());
+    transform(vni_addresses.begin(), vni_addresses.end(), std::back_inserter(addresses),
+               [](const auto& kv) { return kv.first; });
+
+    return false;
+}
+
+void DashPaValidationOrch::paValidationConfigProcess(const PaValidationEntryList& toCreate, const PaValidationEntryList& toRemove)
+{
+    SWSS_LOG_ENTER();
+
+    auto task_status = paValidationConfigApply(toCreate, toRemove);
+    if (task_status != task_success)
+    {
+        parseHandleSaiStatusFailure(task_status);
+    }
+}
+
+task_process_status DashPaValidationOrch::paValidationConfigApply(const PaValidationEntryList& toCreate, const PaValidationEntryList& toRemove)
+{
+    EntityBulker<sai_dash_pa_validation_api_t> bulker(sai_dash_pa_validation_api, gMaxBulkSize);
+    std::vector<sai_status_t> statuses;
+
+    SWSS_LOG_ENTER();
+
+    statuses.reserve(toCreate.size() + toRemove.size());
+
+    for (const auto& entry: toCreate)
+    {
+        bool new_entry = entryRefCountInc(entry);
+        if (!new_entry)
+        {
+            // mark entry as created to omit error handling
+            statuses.emplace_back(SAI_STATUS_SUCCESS);
+        } else {
+            sai_attribute_t attr;
+            uint32_t attr_count = 1;
+
+            sai_pa_validation_entry_t sai_entry;
+            sai_entry.vnet_id = entry.vnet_oid;
+            sai_entry.switch_id = gSwitchId;
+            swss::copy(sai_entry.sip, entry.address);
+
+            attr.id = SAI_PA_VALIDATION_ENTRY_ATTR_ACTION;
+            attr.value.u32 = SAI_PA_VALIDATION_ENTRY_ACTION_PERMIT;
+
+            statuses.emplace_back();
+            bulker.create_entry(&statuses.back(), &sai_entry, attr_count, &attr);
+
+            gCrmOrch->incCrmResUsedCounter(entry.address.isV4() ? CrmResourceType::CRM_DASH_IPV4_PA_VALIDATION : CrmResourceType::CRM_DASH_IPV6_PA_VALIDATION);
```

Comment:
Does this CRM_DASH_IPV4_PA_VALIDATION and CRM_DASH_IPV6_PA_VALIDATION include both outbound and inbound direction entries?

Reply:
Currently, there is no distinction between inbound/outbound PA Validation entries on SAI/DASH level.